update CoffeeScript loader for incoming hook consolidation

JakobJingleheimer opened the pull request.

JakobJingleheimer committed:
update CoffeeScript loader for incoming hook consolidation

JakobJingleheimer commented:
@GeoffreyBooth I think the trick of appending `.js` coffeescript-loader was using previously is no-longer possible with `getFormat()` and `getSource()` consolidated because the same `url` is used for both internally (whereas previously, the appended url was supplied to only getFormat, so getSource still worked). `load()` fails when it subsequently reaches the getSource step because the fake `.coffee.js` truly does not exist.

JakobJingleheimer committed:
add getPackageType util and update coffeescript loader

* importing shout.coffee (cjs) is an empty object

JakobJingleheimer committed:
update HTTPS loader for incoming hook consolidation

JakobJingleheimer committed:
remove debugging code from CoffeeScript loader et al & update bad test case

JakobJingleheimer committed:
consolidate getPackageType into CoffeeScript loader

GeoffreyBooth commented on coffeescript-loader/fixtures/esm-and-commonjs-imports.coffee:
```diff
@@ -4,5 +4,6 @@ import { scream } from './esm/scream.coffee'
 console.log scream 'Hello from ESM'
 
 import shoutDefault from './cjs/shout.coffee'
+console.log shoutDefault # TODO: remove
```
?

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -1,6 +1,15 @@
 import { readFileSync } from 'fs';
 import { createRequire } from 'module';
-import { URL, pathToFileURL } from 'url';
+import {
+  dirname,
+  resolve as resolvePath,
+} from 'path';
+import {
+  fileURLToPath,
+  pathToFileURL,
+  URL,
+} from 'url';
```
This is ultimately going into the docs, so we should follow the style of the other examples. I think (?) they usually write these inline unless they need to wrap.

Also since we’re not importing some other `resolve`, we shouldn’t rename `path`‘s. And I think `URL` is a JavaScript global, not something we need to import.

```suggestion
import { dirname, resolve } from ‘path’;
import { fileURLToPath, pathToFileURL } from ‘url’;
```

JakobJingleheimer replied:
path's `resolve` must be renamed or namespaced to avoid conflicting with the resolve hook, which can't be renamed. I don't have much preference between renaming the import vs namespacing it (the rename seemed DRYer).

👍 for the rest :)

GeoffreyBooth replied:
> path’s `resolve` must be renamed or namespaced to avoid conflicting with the resolve hook, which be renamed. I don't have much preference between renaming the import vs namespacing it (the rename seemed DRYer).
> 
> 👍 for the rest :)

Ah, good point.

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -10,7 +19,7 @@ const baseURL = pathToFileURL(process.cwd() + '/').href;
 const extensionsRegex = /\.coffee$|\.litcoffee$|\.coffee\.md$/;
 
-export function resolve(specifier, context, defaultResolve) {
+export async function resolve(specifier, context, defaultResolve) {
```
This function isn’t actually async. On the one hand I can see the value in copying the signature even if we’re not using every part of it, but on the other this feels wrong.

JakobJingleheimer replied:
I wavered on this a few times, but decided, since this would be an official example, it should have the `async` to remind the user that hooks are async, regardless of whether they use async.

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
+    // CoffeeScript files can be either CommonJS or ES modules, but since load
+    // handles both format and source based on the same url, it cannot be used
+    // for non-js files. Instead, a quick search up the filesystem for a
+    // package.json with a `type` field settles the format issue.
```
This rewritten comment only makes sense to someone who knows the history that there used to be a separate `getFormat` hook. We could phrase it more similarly to the original:

```suggestion
    // CoffeeScript files can be either CommonJS or ES modules, so we want any
    // CoffeeScript file to be treated by Node.js the same as a .js file would
    // be at the same location. To determine how Node.js would interpret an
    // arbitrary .js file, search up the file system for the nearest parent
    // package.json file and read its "type" field.
```

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
+    // handles both format and source based on the same url, it cannot be used
+    // for non-js files. Instead, a quick search up the filesystem for a
+    // package.json with a `type` field settles the format issue.
+    const format = await getPackageType(dirname(fileURLToPath(url)));
```
I feel like the input of `getPackageType` should just be `url`, not the parent folder of `url`. I get that that makes the recursion a little more confusing, but it makes the signature of this function to outsiders a little more obvious (at least to me).

JakobJingleheimer replied:
Actually, yeah, let's DRY this up 😀

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
-  // Let Node.js handle all other URLs.
-  return defaultGetFormat(url, context, defaultGetFormat);
-}
+    // source is ignored (never checked) for cjs, so safe to omit
+    if (format === 'commonjs') return { format };
```
What does your comment here mean? We’re also losing important information in the original comment. Perhaps something like this? And we also always include braces for `if` blocks, I think.

```suggestion
    // If this hook returns a format of 'commonjs', the source will be ignored.
    // To handle CommonJS files, a handler will need to be registered with
    // require.extensions in order to process the files with the CommonJS loader.
    // Avoiding the need for a separate CommonJS handler is a future enhancement
    // planned for ES module loaders.
    if (format === 'commonjs') {
      return { format };
    }
```

JakobJingleheimer replied:
🤔 I've seen many instances of the curly brackets omitted in source, but a quick look thru the docs suggests superfluous curly brackets are present. If you realllllllly want.

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
 
-export function transformSource(source, context, defaultTransformSource) {
-  const { url, format } = context;
+    const { source: rawSource } = await defaultLoad(url, { format });
```
So this works when you run it? `defaultLoad` will return the source of e.g. `file.coffee` with `format: 'module'`?

JakobJingleheimer replied:
Yes, `rawSource` is the actually (string of) CoffeeScript code here. At some point when I was debugging, I logged this, so I saw it 🙂 What's surprising? (That's what we were hoping to facilitate with `getPackageType()`)

GeoffreyBooth replied:
I was surprised that this worked because I remembered there being a check in Node’s default hook that it would always throw on unknown extensions. Maybe that’s in Node’s default `resolve` hook, not `load`? Or we took it out? I can’t remember.

JakobJingleheimer replied:
Yes, that's resolve—resolve is the bouncer at this club 🙂

GeoffreyBooth replied:
Okay, this sounds good. One more question: why are we passing `format` at all? I would think that `defaultLoad` shouldn’t need it, since its signature is the same `source, context, noop`; I’d think that just `defaultLoad(url)` should work here?

JakobJingleheimer replied:
`format` here is necessary because this loader has knowledge that Node does not (it knows to ignore the file extension and look for a package.json): omitting `format` here causes `defaultLoad()` to invoke `defaultGetFormat()` under the hood, which will fail, and thus abort before invoking `defaultGetSource()` (`defaultGetSource()` is all we actually want here, but it is not directly accessible).

GeoffreyBooth replied:
I thought we were collapsing the Node default hooks too, so that `defaultLoad` is just one function rather than a wrapper over the two legacy `defaultGetFormat` and `defaultGetSource` functions.

JakobJingleheimer replied:
We could do; I don't recall that specifically being mentioned.

When I was implementing it, keeping them separate and wrapping with defaultLoad seemed cleaner (single responsibility, separation of concerns, etc).

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
+      bare: true,
+      filename: url,
+    });
 
-  // This hook converts CoffeeScript source code into JavaScript source code
-  // for all imported CoffeeScript files.
-  if (extensionsRegex.test(url)) {
     return {
-      source: CoffeeScript.compile(source, {bare: true, filename: url})
+      format,
+      source: transformedSource,
```
I think we try to keep the examples as brief as is reasonable, since long examples intimidate readers. I’m not sure we need to make this section more verbose.

Is `rawSource` not already a string? Why do we need to cast it?

```js
  // This hook converts CoffeeScript source code into JavaScript source code
  // for all imported CoffeeScript files.
  if (extensionsRegex.test(url)) {
    return {
      source: CoffeeScript.compile(`${rawSource}`, {bare: true, filename: url})
```

JakobJingleheimer replied:
`rawSource` (as returned by `defaultLoad()`) is a string

JakobJingleheimer replied:
>I’m not sure we need to make this section more verbose.

Sure. which part would you like removed—the `transformedSource` constant? That seemed to be worthwhile to explicitly call out (plus is a bit cleaner and easier to debug), but I don't have particularly strong feelings about it. But please do let me know if that is the bit you'd like consolidated.

>Is rawSource not already a string? Why do we need to cast it?

That's what `''+` does 🙂 last I checked, it was the most efficient way (by a lot) to cast to a string.

GeoffreyBooth replied:
If you think `transformedSource` makes it clearer, then sure, it’s worth the added lines.

Yes either `''+rawSource` or `` `${rawSource}` `` will cast `val` into a string, but why is either approach necessary? Is `rawSource` not already a string?

JakobJingleheimer replied:
Because it _can_ also be multiple kinds of Array: [docs/ESM → getSource](https://nodejs.org/api/esm.html#esm_getsource_url_context_defaultgetsource) (the return types haven't changed)

GeoffreyBooth replied:
I’m wondering when it _wouldn’t_ be a string; maybe only for wasm? But I guess for an example it’s better to do `` `${rawSource}` `` or `rawSource.toString()` to be really explicit and hint that the loader author shouldn’t assume that `rawSource` is necessarily always a string.

JakobJingleheimer replied:
👍 explicit is better 🙂

As for when, I'm not sure of the conditions off the top of my head. I do know that when I was debugging the import CJS issue, some `source`s were buffers and not strings (and some were strings). I remember the CoffeeScript lib import was a string, and I'm ~80% sure this loader's test fixture was a buffer. If you'd like to know for sure, I can add back the logging and check in the morning.

GeoffreyBooth replied:
> If you’d like to know for sure, I can add back the logging and check in the morning.

No, please don’t spend any more time on this, it’s not worth the effort. I remember Bradley saying something about the source of a wasm import always being a buffer, so I bet it has to do with that (at least).

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
+  return defaultLoad(url, context, defaultLoad);
+}
+
+function getPackageType(dir) {
+  const packagePath = resolvePath(dir, 'package.json');
+
+  return readFile(packagePath, { encoding: 'utf8' })
+    .then((filestring) => JSON.parse(filestring))
+    .then(({ type }) => type)
+    .catch((err) => {
+      if (err?.code !== 'ENOENT') console.error(err);
+
+      return dir.length > 1 && getPackageType(resolvePath(dir, '..'))
+    });
```
Why not just call `readFile` in a `try`/`catch` block, and save the filesystem call? We should also write this in `async`/`await` style. Something like (untested):

```suggestion
function getPackageType(url) {
  const containingFolder = dirname(fileURLToPath(url));
  try {
    const packageMetadata = await readFile(join(containingFolder, 'package.json'), 'utf-8');
  } catch {
    if (containingFolder.length > 1) {
      return getPackageType(join(containingFolder, '..', 'package.json'));
    } else {
      // We’ve reached the root of the file system and didn’t find a package.json
      return 'commonjs';
    }
  }
  return JSON.parse(packageMetadata).type || 'commonjs';
```

JakobJingleheimer replied:
Hmm? How does a `try`/`catch` avoid an FS call? If you're thinking there should be a check on `containingFolder.length`, it already does that:

```js
// .catch
return dir.length > 1 && getPackageType(resolvePath(dir, '..'))
```

If `dir.length > 1` fails, it short-circuits and settles with boolean `false` (instead of recursing again, effectively avoiding the predestined failing fs call I think you're concerned about). I did that intentionally so it would fail a subsequent typecheck (as well as a truthy/falsy check), causing Node to alert the user that something has gone wrong (whereas returning a valid value could effectively silence a problem, where the user is expecting their value to be found). I would ideally settle with `undefined` (since literally no value was defined), but that's more verbose to implement, so I opted for succinct-and-close-enough.

Try / catch is ugly and awful (and forces more nesting). I would only await when I'm fairly certain whatever I'm awaiting will not throw (and something above catches it), or i would attach a `.catch` to the awaited call. I'm not so keen to provide an example I myself would reject in a PR. I also find returning the promise chain to be much cleaner—it clearly says "the final result of this will be the output" (whereas try/catch can have difficult-to-notice early returns, and deeply nested early returns are especially difficult to notice for a poorly sighted person).

If you feel realllly strongly about the try/catch 😬

GeoffreyBooth replied:
A few things: the filesystem call I was thinking of was `resolvePath`; does that not hit the filesystem? I’m not that familiar with `path.resolve`, so I’d believe you if you told me it didn’t. I’m pretty sure that `path.join` just does string concatenation without any file I/O, which is why I was preferring it.

As for `try`/`catch`, I don’t feel strongly about that; though I’m not sure why `.catch` is significantly better? I’m just thinking we should follow the style of the docs, and I think the newest examples in the docs try to generally use `await` as much as possible instead of the older Promise chain or callback styles. If there’s a way to use `await` without `try`/`catch`, sure, do that; otherwise, why would you prefer a promise chain over `await` with `try`/`catch`? At the very least I think you should collapse the two `.then` calls, since the `JSON.parse` isn’t returning a promise: `.then(fileContents => JSON.parse(fileContents).type || 'commonjs')`.

You also need to account for the case where you find a `package.json` but it doesn’t have a `type` field, which is treated as if it had `"type": "commonjs"`.

JakobJingleheimer replied:
As I understood it, the only difference between `join` and `resolve` is that `resolve` will prepend the normalised path with the cwd when the result is otherwise not an absolute path (whereas `join` would just return the unprepended path segment), and I think the prepending behaviour is desirable here juuust in case (if `path.join` resulted in a non-absolute path, I think the correct version of that would be the cwd-prepended value that `path.resolve` does).

I'm not sure if `process.cwd()` is actually an fs call (I would guess Node would cache this and update its cache on change).

Aside from that, I just checked both the [win32](https://github.com/nodejs/node/blob/master/lib/path.js#L158) and [posix implementations](https://github.com/nodejs/node/blob/master/lib/path.js#L1085), and `path.resolve` looks to be merely string manipulation 🙂

> At the very least I think you should collapse the two `.then` calls, since the `JSON.parse` isn’t returning a promise

Ahh yes. I was thinking of fetch's `rsp.json()`, which does return a promise.

> You also need to account for the case where you find a `package.json` but it doesn’t have a `type` field, which is treated as if it had `"type": "commonjs"`

Ahh, yes. Thanks!

GeoffreyBooth commented on coffeescript-loader/loader.js:
```diff
@@ -71,6 +89,8 @@ const require = createRequire(import.meta.url);
   require.extensions[extension] = (module, filename) => {
     const source = readFileSync(filename, 'utf8');
-    return CoffeeScript.compile(source, {bare: true, filename});
+    const transformedSource = CoffeeScript.compile(source, { bare: true, filename });
+
+    return transformedSource
```
I guess you didn’t think it was obvious that `CoffeeScript.compile` was returning the transformed source?

JakobJingleheimer replied:
Oh 😅 actually that was from when I was debugging the failing cjs coffeescript thing (i wanted to console.log the output of compile to ensure it was proper—it is). I can put it back :)

GeoffreyBooth commented on coffeescript-loader/test.js:
```diff
@@ -1,26 +1,32 @@
   '--experimental-loader',
-  './loader.js',
-  './fixtures/esm-and-commonjs-imports.coffee'
+  fileURLToPath(new URL('./loader.js', import.meta.url).href),
+  fileURLToPath(new URL('./fixtures/esm-and-commonjs-imports.coffee', import.meta.url).href),
```
Why are these changes necessary?

JakobJingleheimer replied:
Yes, very 😉

GeoffreyBooth replied:
I don’t understand; are you saying that these changes were accidental?

JakobJingleheimer replied:
Oh, sorry you asked "why". This is a bit of an edge-case, but when the cwd is not the node-loaders project root, this fails.

For instance, in my local:

```
~/Workspace/
  …/
    nodejs/
    node-loaders/
```

I cd into the nodejs dir, which has my pr-compiled node executable, and run:

```console
~/Workspace/…/nodejs $> ./node \
--experimental-loader \
../node-loaders/coffeescript-loader/loader.js \
../node-loaders/coffeescript-loader/test.js
```

Without that extra bit of code, the files are not found.

GeoffreyBooth replied:
Okay, that’s fine then. I think the most likely real-world case would be where the loader is in `node_modules` for the current project. Like say I add `loader.js` to CoffeeScript, then you’d run something like:

```shell
node --experimental-loader ./node_modules/coffeescript/loader.js app.coffee
```

So whatever we put here should work for this case as well.

JakobJingleheimer replied:
Mm, I think the rule for this would be:

* refs to things internal to the loader module would need to be relative to the loader module (eg what you see above with `import.meta.url`)
* refs to things external to the loader module would need to be relative to the caller (`process.cwd()`).

STRd6 replied:
This breaks the tests when running on Windows with GitBash.

GeoffreyBooth replied:
Since I wrote that code, I’ve learned that the arguments here should be URLs, not paths. That might explain why they were broken in Windows.

GeoffreyBooth commented on coffeescript-loader/test.js:
```diff
@@ -1,26 +1,32 @@
   ok(stdout.includes('HELLO FROM ESM'), 'ESM import transpiles');
-  ok(stdout.includes('Hello from CommonJS!'), 'CommonJS import transpiles');
+
+  // this relies on a named export from CommonJS, which is not supposed to work
```
So is this a TODO? Could we spell out more clearly what is supposed to happen, and mention that this is a bug/unimplemented feature?

JakobJingleheimer replied:
I'm honestly not sure what is supposed to happen—I expect this to work now and before. But it doesn't work then or now. I'm afraid to open a can of worms investigating this, but if it's supposed to work, I (biting my tongue so hard) think it should maybe block nodejs/node#37468 (depending on how pervasive the usage of this is).

What currently happens is the `require.extension` works, the transformed source disappears into the CJS loader (I confirmed it does get passed it), and then an empty object comes out the other side.

GeoffreyBooth replied:
There’s no need to block a PR because it doesn’t fix an existing bug, especially not for an experimental feature. I think all we need to do at this point is add a comment here explaining something like “known issue: blah blah” with the clearest explanation you can write, and we’ll come back to it later. I just don’t want to forget about this or have this code in here and someone like you comes along and thinks that it should be working, or that it did work in some older version of Node but got broken somehow more recently.

GeoffreyBooth commented on https-loader/loader.js:
```diff
@@ -20,31 +20,23 @@ export function resolve(specifier, context, defaultResolve) {
         res.on('data', chunk => data += chunk);
-        res.on('end', () => resolve({ source: data }));
+        res.on('end', () => resolve({
+          // assume all network-provided JavaScript is ES module code
```
```suggestion
          // This example assumes all network-provided JavaScript is ES module code.
```

The “This example” part is important so that readers know that it’s an assumption made because this is an example, not because it’s generally a safe assumption to make.

GeoffreyBooth commented on https-loader/test.js:
```diff
@@ -19,5 +20,6 @@ child.stdout.on('data', (data) => {
 });
 
 child.on('close', (code, signal) => {
+  console.log(`stdout "${stdout}"`)
```
?

JakobJingleheimer replied:
Weird…I remember removing these console logs…

JakobJingleheimer committed:
🧹 remove debugging code

JakobJingleheimer committed:
tweak CoffeeScript Loader dev note

JakobJingleheimer committed:
DRY up getPackageType() to accept a `url` instead of directory

* formatting

JakobJingleheimer committed:
🧹 tidy dev note and code styles

JakobJingleheimer committed:
🧹 debug code cleanup

JakobJingleheimer committed:
🧹 tidy debug code & dev comments

JakobJingleheimer committed:
consolidate json parsing and always potentially invoke next recursion round

* handle package.json files without "type" field

JakobJingleheimer commented:
Okay, so I did some debugging, and I think we have some kind of red-herring here: the cjs deps from coffeescript files do not hit ESMLoader at all. I added logging to `ESMLoader::getModuleJob()` (which all imports handled by ESMLoader go through), and it does not log anything for files imported by `./fixtures/esm-and-commonjs-imports.coffee`. The only commonjs import it handles is coffeescript itself.

To be doubly sure I wasn't missing anything, I added copious logging to module translators' commonjs strategy, and there is nothing whatsoever for files imported by `./fixtures/esm-and-commonjs-imports.coffee`.

So I think we can confidently say whatever is happening is not ESMLoader's fault.

JakobJingleheimer committed:
doc: add caveat note RE importing CJS named exports

GeoffreyBooth commented:
> So I think we can confidently say whatever is happening is not ESMLoader’s fault.

That’s fine, but we still need _some_ way for users to write a loader that affects both ESM and CommonJS files. Like what I was trying to do with the CoffeeScript loader defining `resolve` and `load` but also `require.extensions`. It doesn’t need to be streamlined, like we can have people use the old `require.extensions` still if we must or if that’s the easiest way (at least for now) but there should be some way to have a loader that works for a project that has mixed types. I’d imagine that this might require changes to the CommonJS loader, which means we should absolutely come back to it later as a separate PR, but it would be good to put a comment here saying that we’re aware of the issue. Something like “`TODO: Known limitation: currently, sources returned with format: 'commonjs' are ignored. This will be addressed in a future update.`”

To take another example, say you wanted to create a YAML loader. As a user of such a loader, I would expect that both `import data from 'data.yaml'` and `const data = require('data.yaml')` should work when I have the loader enabled. Ditto for CoffeeScript or TypeScript or other more sophisticated file types that can have their own imports and are more JavaScript-like.

JakobJingleheimer commented:
From the debugging I did, I did not see any way to get it to work as things currently are, but it may have just needed more investigation and/or more knowledge of the CJSLoader.

Regarding (future) updates to the CJSLoader, that sounds likely needed, but I remember someone (Bradley?) saying 2–3 team meetings ago that something like this would require changes to the CJSLoader, but previous efforts failed in a fiery blaze because there was too much disagreement.

I would certainly like it to be fixed, but if that's true, maybe we shouldn't promise the issue will be addressed?

GeoffreyBooth commented:
> Regarding (future) updates to the CJSLoader, that sounds likely needed, but I remember someone (Bradley?) saying 2–3 team meetings ago that something like this would require changes to the CJSLoader, but previous efforts failed in a fiery blaze because there was too much disagreement.
> 
> I would certainly like it to be fixed, but if that’s true, maybe we shouldn’t promise the issue will be addressed?

I think a comment can make whatever promises it wants. No one will hold that comment responsible 😄 

To the broader point, though, I think we should tap Bradley (and Guy, if he has the time) to get their wisdom about what the issue is with supporting `format: 'commonjs'` and what kind of refactoring would be required to get that to work. Because it not working is pretty plainly a bug, and bugs must be fixed, or we change the API to remove `format: 'commonjs'` as one of the options. It can’t just be left broken because of disagreements; that doesn’t fly, and I think I could get the TSC to back me up on that one if necessary (but I doubt that would be necessary).

Anyway I think let’s add a comment and move on for now and get this PR landed, and after that maybe we can create an issue on nodejs/loaders spelling out the problem as detailed as we can, and one of the others can hopefully tell us where to look in the code for what needs to be changed to fix it.

JakobJingleheimer commented:
Sounds good to me :)

I added such a comment to the CoffeeScript loader test file (lines 30–35). Does that suffice, or were you thinking to add the note elsewhere more conspicuous (like the ESM docs)?

JakobJingleheimer commented:
Ah, okay, so I got the download from Bradley. The gist of it is what we thought: this part of the two APIs are not integrated. It is possible for the user to facilitate the integration by returning `format: "module"` instead of its actual "commonjs", parsing the source, and manually hooking it up to the exports, which doesn't sound pleasant but also doesn't sound like the worst thing.

@bmeck does that accurately represent what would need to be done from a high level?

JakobJingleheimer committed:
make type coercion more obvious: `''+` → `toString()`

GeoffreyBooth commented:
> The gist of it is what we thought: this part of the two APIs are not integrated. It is possible for the user to facilitate the integration by returning `format: "module"` instead of its actual “commonjs”, parsing the source, and manually hooking it up to the exports, which doesn’t sound pleasant but also doesn’t sound like the worst thing.

I think wherever the docs currently discuss `format` we should add a note. I don’t think we should add an example of how to do what Bradley is describing, since it’s clearly not the intended API, but I think we should document that for now `format: 'commonjs'` doesn’t work as intended or expected.

bmeck commented:
@GeoffreyBooth to be clear it works as expected by my view XD. CJS hooks don't cleanly integrate against stuff.
